File: 02_DB_Interpolation/smallwing/matlab_code/B_read_int_data.m
%% read interpolated data

clc
clear
close all
addpath matlab_func/

%% import mesh

load('int_mesh.mat')

%%

Nx=length(int_mesh.xc);
Ny=length(int_mesh.yn);
Nz=length(int_mesh.z);

Npoints=Nx*Ny*Nz;

sinalpha=sin(int_mesh.alpha);
cosalpha=cos(int_mesh.alpha);

PATH_INT_FIELDS='../run_interpolation/';

N_FILES=2;

IF_SUCTION_SIDE=0;
IF_PRESSURE_SIDE=0;
IF_VELOCITY=1;
IF_VORTICITY=0;
IF_LA2=0;

%% LOAD INPUT
disp('LOAD INPUT')
for i_file=1:N_FILES
%Load file containing number of wall faces per core
fHistoryname         = strcat(PATH_INT_FIELDS,'intHistory.f',num2str(i_file,'%12.5d'));
% fHistoryname         = strcat('/cfs/klemming/nobackup/a/atzori/MADRID/tDuct_Test/run_360/intHistory_duct0.f',num2str(i_file,'%12.5d'));
history = load(fHistoryname);
ncores=history(end,1)+1; % number of cores

disp(ncores)
disp(i_file)

if IF_VELOCITY==1
variablename='intVX.f';
bb_buff_vx=load_int_file(PATH_INT_FIELDS, variablename, i_file, history, ncores);

variablename='intVY.f';
bb_buff_vy=load_int_file(PATH_INT_FIELDS, variablename, i_file, history, ncores);

variablename='intVZ.f';
bb_buff_vz=load_int_file(PATH_INT_FIELDS, variablename, i_file, history, ncores);
end

if IF_LA2==1
variablename='intLA.f';
bb_buff_la=load_int_file(PATH_INT_FIELDS, variablename, i_file, history, ncores);
end

if IF_VORTICITY==1
variablename='intOX.f';
bb_buff_ox=load_int_file(PATH_INT_FIELDS, variablename, i_file, history, ncores);

variablename='intOY.f';
bb_buff_oy=load_int_file(PATH_INT_FIELDS, variablename, i_file, history, ncores);

variablename='intOZ.f';
bb_buff_oz=load_int_file(PATH_INT_FIELDS, variablename, i_file, history, ncores);
end

%% ROTATE

if IF_VELOCITY==1
test_vx=reshape(bb_buff_vx,Nx,Ny,Nz);
test_vy=reshape(bb_buff_vy,Nx,Ny,Nz);
end

if IF_VORTICITY==1
test_ox=reshape(bb_buff_ox,Nx,Ny,Nz);
test_oy=reshape(bb_buff_oy,Nx,Ny,Nz);
end

if IF_SUCTION_SIDE==1
    for i_k=1:Nz
        for i_y=1:Ny
            for i_x=1:Nx

                if IF_VELOCITY==1
                    vx=test_vx(i_x,i_y,i_k);
                    vy=test_vy(i_x,i_y,i_k);
                    test_vx(i_x,i_y,i_k)=vx*cosalpha(i_x)+vy*sinalpha(i_x);
                    test_vy(i_x,i_y,i_k)=vy*cosalpha(i_x)-vx*sinalpha(i_x);
                end

                if IF_VORTICITY==1
                    ox=test_ox(i_x,i_y,i_k);
                    oy=test_oy(i_x,i_y,i_k);
                    test_ox(i_x,i_y,i_k)=ox*cosalpha(i_x)+oy*sinalpha(i_x);
                    test_oy(i_x,i_y,i_k)=oy*cosalpha(i_x)-ox*sinalpha(i_x);
                end

            end
        end
    end
end

if IF_PRESSURE_SIDE==1
    for i_k=1:Nz
        for i_y=1:Ny
            for i_x=1:Nx


                if IF_VELOCITY==1
                    vx=test_vx(i_x,i_y,i_k);
                    vy=test_vy(i_x,i_y,i_k);
                    test_vx(i_x,i_y,i_k)=vx*cosalpha(i_x)+vy*sinalpha(i_x);
                    test_vy(i_x,i_y,i_k)=vx*sinalpha(i_x)-vy*cosalpha(i_x);
                end

                if IF_VORTICITY==1
                    vx=test_vx(i_x,i_y,i_k);
                    vy=test_vy(i_x,i_y,i_k);
                    test_vx(i_x,i_y,i_k)=vx*cosalpha(i_x)+vy*sinalpha(i_x);
                    test_vy(i_x,i_y,i_k)=vx*sinalpha(i_x)-vy*cosalpha(i_x);
                end

            end
        end
    end
end

if IF_VELOCITY==1
    bb_buff_vx=reshape(test_vx,Nx*Ny*Nz,1,1);
    bb_buff_vy=reshape(test_vy,Nx*Ny*Nz,1,1);
end

if IF_VORTICITY==1
    bb_buff_ox=reshape(test_ox,Nx*Ny*Nz,1,1);
    bb_buff_oy=reshape(test_oy,Nx*Ny*Nz,1,1);
end
end

test_vx=reshape(bb_buff_vx,Ny,Nx,Nz);
test_vy=reshape(bb_buff_vy,Ny,Nx,Nz);
test_vz=reshape(bb_buff_vz,Ny,Nx,Nz);

[XX,YY]=meshgrid(int_mesh.xc,int_mesh.yn);

%%

figure
scatter(int_mesh.x_pts,int_mesh.y_pts); hold on; axis tight equal

%%

set(0,'defaultTextInterpreter','latex'); %trying to set the default
lwidth=4; % line width
legend_on=0;  % show the legend
title_on=0;   % show the title
linewidth_axis=2;  % width of the axis line
scatter_with=1;    % width of the scatter line
figx=1200; % figure horizontal size
figy=500; % figure vertical size
font_size=28;      % font size
set(0, 'DefaultAxesFontWeight', 'normal', ...
    'DefaultAxesFontSize', font_size, ...
    'DefaultAxesFontAngle', 'normal', ... % Not sure the difference here
    'DefaultAxesFontWeight', 'normal', ... % Not sure the difference here
    'DefaultAxesTitleFontWeight', 'normal', ...
    'DefaultAxesTitleFontSizeMultiplier', 1) ;

figure('rend','painters','pos',[10 10 figx figy]);
h=pcolor(XX,YY,test_vx(:,:));
set(h, 'EdgeAlpha', 0.1);
axis equal tight
set(gca,'TickLabelInterpreter', 'latex');
title('$u_t=U_t+u_t''$','Interpreter','latex')
xlabel('$x/c$','Interpreter','latex')
ylabel('$y/c$','Interpreter','latex')

figure('rend','painters','pos',[10 10 figx figy]);
h=pcolor(XX,YY,test_vy(:,:));
set(h, 'EdgeAlpha', 0.1);
axis equal tight
set(gca,'TickLabelInterpreter', 'latex');
title('$v_n=V_n+v_n''$','Interpreter','latex')
xlabel('$x/c$','Interpreter','latex')
ylabel('$y/c$','Interpreter','latex')

figure('rend','painters','pos',[10 10 figx figy]);
h=pcolor(XX,YY,test_vz(:,:));
set(h, 'EdgeAlpha', 0.1);
axis equal tight
set(gca,'TickLabelInterpreter', 'latex');
title('$w=W+w''$','Interpreter','latex')
xlabel('$x/c$','Interpreter','latex')
ylabel('$y/c$','Interpreter','latex')

File: 02_DB_Interpolation/NACA4412_5deg/3_convert_to_hdf5/B_read_int_data.py
import h5py
import matplotlib.pyplot as plt
import numpy as np
import os
import scipy.io as sio

from load_int_file import load_int_file

#------------------------------------------------------------------------
# Parameters
#------------------------------------------------------------------------
case_type:str = "suction" # pressure, suction, wake
out_start_index: int = 1  # Index of the first file in the global database
in_start_index: int  = 1  # File to start processing (in case you do it by batches), normally 1
n_files: int = 10000      # Number of files to transform to hdf5 (total in the folder)

# path_int_fields: str = "../2_run_interpolation/INTERP"
path_int_fields: str = "../../../NACA4412_5deg_DB/2_INTERPOLATED_RAW_SS/00001_10000"
path_db_fields: str  = "./hdf5_files"

flag_suction_side: bool  = True
flag_pressure_side: bool = False
flag_velocity: bool      = True
flag_vorticity: bool     = True
flag_la2: bool           = False
flag_project: bool       = False
flag_save_h5: bool       = True

#------------------------------------------------------------------------
# Import the interpolation grid
#------------------------------------------------------------------------
print("Imported interpolation grid:", flush=True)
load_mat: dict = sio.loadmat(f"../0_generate_grid/int_mesh_{case_type}.mat")
int_mesh = load_mat["int_mesh"]

Rec = float(np.squeeze(int_mesh["Rec"][0,0]))
utz = float(np.squeeze(int_mesh["utz"][0,0]))
nu  = 1 / Rec
print(f"{Rec = }")
print(f"{nu = }")
print(f"{utz = }")

x_pts   = np.squeeze(int_mesh["x_pts"][0,0])
y_pts   = np.squeeze(int_mesh["y_pts"][0,0])
z_pts   = np.squeeze(int_mesh["z_pts"][0,0])
alpha   = np.squeeze(int_mesh["alpha"][0,0])
xc      = np.squeeze(int_mesh["xc"][0,0])     # Chord-wise distribution
xs      = np.squeeze(int_mesh["xs"][0,0])     # Wall-tangent distribution
yn      = np.squeeze(int_mesh["yn"][0,0])     # Wall-normal distribution
z       = np.squeeze(int_mesh["z"][0,0])      # Spanwise distribution

Nx      = len(xc)
Ny      = len(yn)
Nz      = len(z)
Npoints = Nx * Ny * Nz
print(f"{Ny = }, {Nx = }, {Nz = }", flush=True)

sinalpha = np.sin(alpha)
cosalpha = np.cos(alpha)

#------------------------------------------------------------------------
# Load the fields
#------------------------------------------------------------------------
# Save the grid information
if flag_save_h5:
    if not os.path.exists(path_db_fields):
        os.makedirs(path_db_fields)
    _fileName: str       = f"naca4412_5deg_200k_{case_type}.grid.h5"
    fileName_grid: str = path_db_fields + "/" + _fileName
    with h5py.File(fileName_grid,'w') as fGrid:
        fGrid.create_dataset("Rec", data=Rec)
        fGrid.create_dataset("utz", data=utz)
        fGrid.create_dataset("xc_vec", data=xc)     # Points in x in the chord-wise reference frame
        fGrid.create_dataset("xs_vec", data=xs)     # Points in x in the wall reference frame
        fGrid.create_dataset("yn_vec", data=yn)     # Points in y in the wall reference frame
        fGrid.create_dataset("z_vec",  data=z)      # Points in z in the wall reference frame       
        fGrid.create_dataset("alpha",  data=alpha)
        fGrid.create_dataset("x_pts",  data=x_pts)
        fGrid.create_dataset("y_pts",  data=y_pts)
        fGrid.create_dataset("z_pts",  data=z_pts)
    print("* Saved grid information", flush=True)

print("")
print("Loading input...", flush=True)
for i_file in range(in_start_index,n_files+1):
    in_index = i_file
    fHistoryname: str        = path_int_fields + "/" + "intHistory.f" + f"{in_index:05d}"
    history: np.ndarray[int] = np.loadtxt(fHistoryname, dtype=int)
    ncores: int              = history[-1,0] + 1
    print(f"- File {i_file:05d}, {ncores = }")

    if flag_velocity:
        variable_name: str = "intVX.f"
        bb_buff_vx, time = load_int_file(path_int_fields, variable_name, i_file, history, ncores)

        variable_name: str = "intVY.f"
        bb_buff_vy, _ = load_int_file(path_int_fields, variable_name, i_file, history, ncores)

        variable_name: str = "intVZ.f"
        bb_buff_vz, _ = load_int_file(path_int_fields, variable_name, i_file, history, ncores)

        int_vx = np.reshape(bb_buff_vx,(Ny,Nx,Nz), order='F')
        int_vy = np.reshape(bb_buff_vy,(Ny,Nx,Nz), order='F')
        int_vz = np.reshape(bb_buff_vz,(Ny,Nx,Nz), order='F')

    if flag_vorticity:
        variable_name: str = "intOX.f"
        bb_buff_ox, time = load_int_file(path_int_fields, variable_name, i_file, history, ncores)

        variable_name: str = "intOY.f"
        bb_buff_oy, _ = load_int_file(path_int_fields, variable_name, i_file, history, ncores)

        variable_name: str = "intOZ.f"
        bb_buff_oz, _ = load_int_file(path_int_fields, variable_name, i_file, history, ncores)

        int_ox = np.reshape(bb_buff_ox,(Ny,Nx,Nz), order='F')
        int_oy = np.reshape(bb_buff_oy,(Ny,Nx,Nz), order='F')
        int_oz = np.reshape(bb_buff_oz,(Ny,Nx,Nz), order='F')

    if flag_la2:
        variable_name: str = "intLA.f"
        bb_buff_la, time = load_int_file(path_int_fields, variable_name, i_file, history, ncores)

        int_la = np.reshape(bb_buff_la,(Ny,Nx,Nz), order='F')       

    #------------------------------------------------------------------------
    # Project the xy vectors to the wall reference frame
    #------------------------------------------------------------------------
    if flag_project:
        cos = cosalpha.reshape(1,-1,1)
        sin = sinalpha.reshape(1,-1,1)
        
        if flag_suction_side:
            if flag_velocity:
                _proj_vx = int_vx * cos + int_vy * sin
                _proj_vy = int_vy * cos - int_vx * sin
                int_vx   = _proj_vx
                int_vy   = _proj_vy
            if flag_vorticity:
                _proj_ox = int_ox * cos + int_oy * sin
                _proj_oy = int_oy * cos - int_ox * sin
                int_ox   = _proj_ox
                int_oy   = _proj_oy
        
        if flag_pressure_side:
            if flag_velocity:
                _proj_vx = int_vx * cos  + int_vy * sin
                _proj_vy = int_vx * sin  - int_vy * cos
                int_vx   = _proj_vx
                int_vy   = _proj_vy
            if flag_vorticity:
                _proj_ox = int_ox * cos  + int_oy * sin
                _proj_oy = int_ox * sin  - int_oy * cos
                int_ox   = _proj_ox
                int_oy   = _proj_oy

    #------------------------------------------------------------------------
    # Save in .h5 format
    #------------------------------------------------------------------------
    if flag_save_h5:
        out_index = out_start_index + i_file - 1
        if flag_velocity:  # Velocity vector
            _fileName: str       = f"naca4412_5deg_200k_{case_type}.{out_index:05d}.h5.uvw"
            fileName_interp: str = path_db_fields + "/" + _fileName
            with h5py.File(fileName_interp,'w') as fInterp:
                fInterp.create_dataset("time", data=time)
                fInterp.create_dataset("vx", data=int_vx)
                fInterp.create_dataset("vy", data=int_vy)
                fInterp.create_dataset("vz", data=int_vz)

        if flag_vorticity:  # Vorticity vector
            _fileName: str       = f"naca4412_5deg_200k_{case_type}.{out_index:05d}.h5.omega"
            fileName_interp: str = path_db_fields + "/" + _fileName
            with h5py.File(fileName_interp,'w') as fInterp:
                fInterp.create_dataset("time", data=time)
                fInterp.create_dataset("ox", data=int_ox)
                fInterp.create_dataset("oy", data=int_oy)
                fInterp.create_dataset("oz", data=int_oz)

        if flag_la2:  # Lambda2 scalar field
            _fileName: str       = f"naca4412_5deg_200k_{case_type}.{out_index:05d}.h5.la2"
            fileName_interp: str = path_db_fields + "/" + _fileName
            with h5py.File(fileName_interp,'w') as fInterp:
                fInterp.create_dataset("time", data=time)
                fInterp.create_dataset("la2", data=int_la)

    print(f"  * Saved {_fileName}, time: {time:.3f}s", flush=True)

#------------------------------------------------------------------------
# Plot interpolation
#------------------------------------------------------------------------
XC, YN = np.meshgrid(xc, yn)

plt.rcParams.update({'font.size': 28})

# # Interpolation grid
# fig = plt.figure(figsize=(12,5))
# plt.scatter(x_pts, y_pts, c=int_vx.T, marker=".", cmap="turbo")
# plt.xlabel("x/c")
# plt.ylabel("y/c")
# plt.axis('scaled')
# plt.tight_layout()
# fig.savefig("grid_u.png",dpi=400)

# Velocity field
fig = plt.figure(figsize=(12,5))
h = plt.pcolormesh(XC, YN, int_vx[:,:,0], edgecolors=("k", 0.5), cmap="turbo")#, vmin=0, vmax=1.4)
plt.xlabel("x/c")
plt.ylabel("n/c")
plt.title("$u_t=U_t+u_t'$")
plt.axis('scaled')
plt.tight_layout()
fig.savefig("u_grid.png",dpi=400)

fig = plt.figure(figsize=(12,5))
h   = plt.pcolormesh(XC, YN, int_vx[:,:,0], edgecolors=("k", 0), cmap="turbo")#, vmin=0, vmax=1.4)
cb  = plt.colorbar(h)
plt.xlabel("x/c")
plt.ylabel("n/c")
plt.title("$u_t=U_t+u_t'$")
plt.axis('scaled')
plt.tight_layout()
fig.savefig("u.png",dpi=400)

fig = plt.figure(figsize=(12,5))
h   = plt.pcolormesh(XC, YN, int_vy[:,:,0], edgecolors=("k", 0), cmap="turbo")#, vmin=-0.36, vmax=0.2)
cb  = plt.colorbar(h)
plt.xlabel("x/c")
plt.ylabel("n/c")
plt.title("$v_t=V_t+v_t'$")
plt.axis('scaled')
plt.tight_layout()
fig.savefig("v.png",dpi=400)

fig = plt.figure(figsize=(12,5))
h   = plt.pcolormesh(XC, YN, int_vz[:,:,0], edgecolors=("k", 0), cmap="turbo")#, vmin=-0.31, vmax=0.29)
cb  = plt.colorbar(h)
plt.xlabel("x/c")
plt.ylabel("n/c")
plt.title("$w_t=W_t+w_t'$")
plt.axis('scaled')
plt.tight_layout()
fig.savefig("w.png",dpi=400)